src/Extension/AccountReset/SecurityExtension.php

```diff
@@ -0,0 +1,204 @@
+            );
+            return $this->owner->getResponse()->setBody($output)->setStatusCode(400);
+        }
+
+        $request->getSession()->set('MemberID', $member->ID);
+
+        return $this->owner->getResponse()->setBody($this->owner->renderWith(
+            'Security',
+            [
+                'Title' => _t(
+                    __CLASS__ . '.ACCOUNT_RESET_TITLE',
+                    'Reset Account'
+                ),
+                'Message' => _t(
+                    __CLASS__ . '.ACCOUNT_RESET_DESCRIPTION',
+                    'Your password will be changed, and any registered MFA methods will be removed.'
+                ),
+                'Form' => $this->ResetAccountForm(),
+            ]
+        ));
+    }
+
+    public function ResetAccountForm(): Form
+    {
+        $fields = FieldList::create([
+            PasswordField::create(
+                'NewPassword1',
+                _t(
+                    'SilverStripe\\Security\\Member.NEWPASSWORD',
+                    'New Password'
+                )
+            ),
+            PasswordField::create(
+                'NewPassword2',
+                _t(
+                    'SilverStripe\\Security\\Member.CONFIRMNEWPASSWORD',
+                    'Confirm New Password'
+                )
+            ),
+        ]);
+
+        $actions = FieldList::create([
+            FormAction::create('doResetAccount', 'Reset Account'),
+        ]);
+
+        $validation = RequiredFields::create(['NewPassword1', 'NewPassword2']);
+
+        $form = Form::create($this->owner, 'ResetAccountForm', $fields, $actions, $validation);
+
+        $this->owner->extend('updateResetAccountForm', $form);
+
+        return $form;
+    }
+
+    /**
+     * Resets the user's password, and triggers other account reset procedures
+     *
+     * @param array $data
+     * @param Form $form
+     * @return HTTPResponse
+     */
+    public function doResetAccount(array $data, Form $form): HTTPResponse
+    {
+        $memberID = $this->owner->getRequest()->getSession()->get('MemberID');
+
+        // If the ID isn't in the session, politely assume the session has expired
+        if (!$memberID) {
+            $form->sessionMessage(
+                _t(
+                    __CLASS__ . '.RESETTIMEDOUT',
+                    "The account reset process timed out. Please click the link in the email and try again."
+                ),
+                ValidationResult::TYPE_ERROR
+            );
+
+            return $this->owner->redirectBack();
+        }
+
+        /** @var Member&MemberExtension $member */
+        $member = Member::get()->byID(intval($memberID));
+
+        // Fail if passwords do not match
+        if ($data['NewPassword1'] !== $data['NewPassword2']) {
+            $form->sessionMessage(
+                _t(
+                    'SilverStripe\\Security\\Member.ERRORNEWPASSWORD',
+                    'You have entered your new password differently, try again'
```

Comment:
This renders a little weird in login forms. Might be my fault from my last PR there though...

![image](https://user-images.githubusercontent.com/5170590/59309278-4c836100-8cf7-11e9-9122-a79f8cba021b.png)

Also, are you sure about this wording? It doesn't sit perfectly with me.

Reply:
I think the fix for the styling will live in login-forms, yeah. The wording matches the core Reset Password UI, so it's probably best to leave as-is?